cmd/gateway/s3x/crdt_dagsyncer.go

```diff
@@ -0,0 +1,136 @@
+package s3x
+
+import (
+	"context"
+	"fmt"
+
+	pb "github.com/RTradeLtd/TxPB/v3/go"
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	ipld "github.com/ipfs/go-ipld-format"
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+)
+
+//crdtDAGSyncer implements crdt.DAGSyncer using a pb.NodeAPIClient and a datastore
+type crdtDAGSyncer struct {
+	client pb.NodeAPIClient
+	ds     datastore.Batching
+}
+
+// Get retrieves nodes by CID. Depending on the NodeGetter
+// implementation, this may involve fetching the Node from a remote
+// machine; consider setting a deadline in the context.
+func (d *crdtDAGSyncer) Get(ctx context.Context, c cid.Cid) (ipld.Node, error) {
+	resp, err := d.client.Dag(ctx, &pb.DagRequest{
+		RequestType: pb.DAGREQTYPE_DAG_GET,
+		Hash:        c.String(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	block := blocks.NewBlock(resp.RawData)
+	if block.Cid() != c {
+		return nil, errors.New("unexpected data received from node server")
+	}
+	n, err := ipld.Decode(block)
+	return n, d.setBlock(c, err)
+}
+
+// GetMany returns a channel of NodeOptions given a set of CIDs.
+func (d *crdtDAGSyncer) GetMany(ctx context.Context, cs []cid.Cid) <-chan *ipld.NodeOption {
+	out := make(chan *ipld.NodeOption, len(cs))
+	go func() {
+		for _, c := range cs {
+			n, err := d.Get(ctx, c)
+			out <- &ipld.NodeOption{
+				Node: n,
+				Err:  err,
+			}
+		}
+		close(out)
+	}()
+	return out
+}
+
+// Add adds a node to this DAG.
+func (d *crdtDAGSyncer) Add(ctx context.Context, n ipld.Node) error {
+	return d.AddMany(ctx, []ipld.Node{n})
+}
+
+// AddMany adds many nodes to this DAG.
+//
+// Consider using the Batch NodeAdder (`NewBatch`) if you make
+// extensive use of this function.
+func (d *crdtDAGSyncer) AddMany(ctx context.Context, ns []ipld.Node) error {
+	cs := make([]string, 0, len(ns))
+	for _, n := range ns {
+		links := make(map[string]string, len(n.Links()))
+		for i, l := range n.Links() {
+			name := l.Name
+			if name == "" { // a work around if links don't have names
+				name = fmt.Sprint(i)
+			}
+			links[name] = l.Cid.String()
+		}
+		r, err := d.client.Dag(ctx, &pb.DagRequest{
+			RequestType: pb.DAGREQTYPE_DAG_NEW_NODE,
+			Data:        n.RawData(),
+			Hash:        n.Cid().String(),
+			Links:       links,
+		})
+		if err != nil {
+			return err
+		}
+		if len(r.Hashes) != 1 {
+			return errors.New("unexpected number of hashes returned")
+		}
+		if r.Hashes[0] != n.Cid().String() {
+			return errors.New("unexpected Cid returned")
+		}
+		if err := d.setBlock(n.Cid()); err != nil {
+			return err
+		}
+		cs = append(cs, n.Cid().String())
+	}
+	_, err := d.client.Persist(ctx, &pb.PersistRequest{Cids: cs}) //Question: should persist be requested before add?
+	if err != nil {
+		return err
+	}
+	return nil
+}
```

Comment:
```
--- FAIL: TestS3X_xObjects_GetHash_Crdt (0.46s)

    gateway-s3x_test.go:36: error writing new block QmS5MwH3Px8gGeNtZDKT59BGUeGinUgQhaFbbqw4MumE3D: unexpected Cid returned
```

@bonedaddy I'm probably doing `d.client.Dag(ctx, &pb.DagRequest{` incorrectly.

Reply:
I'll take a look at this today and try to figure out whats up

Reply:
OK I noticed two things, you don't need to set  `Hash:        n.Cid().String(),` since that should only be for retrieval calls. Also this set of lines here:

```
		for i, l := range n.Links() {
			name := l.Name
			if name == "" { // a work around if links don't have names
				name = fmt.Sprint(i)
			}
			links[name] = l.Cid.String()
		}
```

will adjust the expects hashes. The name of links are included in the values used to calculate the CIDs, so here you are altering them, and will be getting a different value. I think that should cover the differences. Also one thing to keep in mind:

```
		if r.Hashes[0] != n.Cid().String() {
			return errors.New("unexpected Cid returned")
		}
```

I would probably change this error to:

`fmt.Errorf("unexpected Cid returned, got %s, want %s", r.Hashes[0], n.Cid().String())` so we can get some more debugging information.

Reply:
Got it to work now, thanks.